src/chrome/internal/scripts/sourcesMapper.ts

```diff
@@ -2,61 +2,133 @@
+import { LocationInLoadedSource, LocationInScript, Position } from '../locations/location';
+import { ILoadedSource } from '../sources/loadedSource';
+import { IResourceIdentifier, parseResourceIdentifier } from '../sources/resourceIdentifier';
+import { IScript } from './script';
 
-export interface ISourcesMapper {
-    readonly sources: string[];
-    getPositionInSource(positionInScript: IPositionInScript): IPositionInSource | null;
-    getPositionInScript(positionInSource: IPositionInSource): IPositionInScript | null;
+export interface ISourceToScriptMapper {
+    getPositionInScript(positionInSource: LocationInLoadedSource): LocationInScript | null;
 }
 
-interface IPositionInSource {
-    readonly source: string;
-    readonly line: LineNumber;
-    readonly column?: ColumnNumber;
+export interface IScriptToSourceMapper {
+    getPositionInSource(positionInScript: LocationInScript): LocationInLoadedSource | null;
 }
 
-interface IPositionInScript {
-    readonly line: LineNumber;
-    readonly column?: ColumnNumber;
+export interface ISourceMapper extends ISourceToScriptMapper, IScriptToSourceMapper { }
+
+export interface IMappedSourcesMapper extends ISourceMapper {
+    readonly sources: string[];
 }
 
 /** This class maps locations from a script into the sources form which it was compiled, and back. */
-export class SourcesMapper implements ISourcesMapper {
-    public getPositionInSource(positionInScript: IPositionInScript): IPositionInSource | null {
-        const mappedPosition = this._sourceMap.authoredPositionFor(positionInScript.line, positionInScript.column || 0);
-        return mappedPosition && mappedPosition.source && mappedPosition.line
-            ? { source: mappedPosition.source, line: createLineNumber(mappedPosition.line), column: createColumnNumber(mappedPosition.column) }
-            : null;
+export class MappedSourcesMapper implements IMappedSourcesMapper {
+    private readonly _rangeInSources: Map<IResourceIdentifier, SourceRange>;
+
+    public getPositionInSource(positionInScript: LocationInScript): LocationInLoadedSource | null {
+        const scriptPositionInResource = this._script.rangeInSource.start.position;
+
+        // All the lines need to be adjusted by the relative position of the script in the resource (in an .html if the script starts in line 20, the first line is 20 rather than 0)
+        const lineNumberRelativeToScript = positionInScript.position.lineNumber - scriptPositionInResource.lineNumber;
+
+        // The columns on the first line need to be adjusted. Columns on all other lines don't need any adjustment.
+        const columnNumberRelativeToScript = (lineNumberRelativeToScript === 0 ? scriptPositionInResource.columnNumber : 0) + (positionInScript.position.columnNumber || 0);
+
+        const mappedPosition = this._sourceMap.authoredPositionFor(lineNumberRelativeToScript, columnNumberRelativeToScript);
+
+        if (mappedPosition && mappedPosition.source && mappedPosition.line) {
```

Comment:
`mappedPosition.line` could be 0 right, dangerous truthiness check?

Reply:
Nice catch